File: AdventOfCode/2018/AdventOfCode/Day18.cs
using System;

namespace AdventOfCode
{
    public static class Day18
    {
        public static int Part1(string input)
        {
            var lines = input.Split("\n".ToCharArray());
            var gridSize = lines.Length;

            var grid = new char[gridSize, gridSize];
            
            for (int i = 0; i < gridSize; i++)
            {
                for (int j = 0; j < gridSize; j++)
                {
                    grid[i, j] = lines[i][j];
                }
            }

            for (int iteration = 0; iteration < 10; iteration++)
            {
                var next = new char[gridSize, gridSize];

                for (int x = 0; x < gridSize; x++)
                {
                    for (int y = 0; y < gridSize; y++)
                    {
                        var numAdjacentTrees = 0;
                        var numAdjacentLumberyards = 0;

                        for (int dx = -1; dx <= 1; dx++)
                        {
                            for (int dy = -1; dy <= 1; dy++)
                            {
                                if ((dx == 0 && dy == 0)
                                    || x + dx < 0
                                    || y + dy < 0
                                    || x + dx >= gridSize
                                    || y + dy >= gridSize)
                                {
                                    continue;
                                }

                                switch (grid[x + dx, y + dy])
                                {
                                    case '|':
                                        numAdjacentTrees++;
                                        break;
                                    case '#':
                                        numAdjacentLumberyards++;
                                        break;
                                }
                            }
                        }

                        char nextValue;

                        switch (grid[x, y])
                        {
                            case '.':
                                nextValue = numAdjacentTrees >= 3 ? '|' : '.';
                                break;
                            case '|':
                                nextValue = numAdjacentLumberyards >= 3 ? '#' : '|';
                                break;
                            case '#':
                                nextValue = numAdjacentLumberyards >= 1 && numAdjacentTrees >= 1 ? '#' : '.';
                                break;
                            default:
                                throw new Exception("Invalid character");
                        }

                        next[x, y] = nextValue;
                    }
                }

                grid = next;
            }

            var numTrees = 0;
            var numLumberyards = 0;

            for (int i = 0; i < gridSize; i++)
            {
                for (int j = 0; j < gridSize; j++)
                {
                    if (grid[i, j] == '|')
                    {
                        numTrees++;
                    }
                    else if (grid[i, j] == '#')
                    {
                        numLumberyards++;
                    }
                }
            }

            return numTrees * numLumberyards;
        }

        public static int Part2(string input)
        {
            return 0;
        }
    }
}

File: AdventOfCode/2018/AdventOfCode/Day18.py
def get_hash(grid):
    result = ""
    for y in range(grid_size):
        for x in range(grid_size):
            result += grid[x][y]

    return result


cur_iteration = [list(line.strip()) for line in open("day18.txt").readlines()]
grid_size = len(cur_iteration)

seen = dict()
iteration = 0
seen[get_hash(cur_iteration)] = iteration

while True:
    next_iteration = [line[:] for line in cur_iteration]

    for y in range(grid_size):
        for x in range(grid_size):
            num_adj_trees = 0
            num_adj_lumberyards = 0

            for dy in range(-1, 2):
                for dx in range(-1, 2):
                    if (dx == 0 and dy == 0) or x + dx < 0 or y + dy < 0 or x + dx >= grid_size or y + dy >= grid_size:
                        continue

                    value = cur_iteration[x + dx][y + dy]
                    if value == "|":
                        num_adj_trees += 1
                    elif value == "#":
                        num_adj_lumberyards += 1

            if cur_iteration[x][y] == ".":
                next_iteration[x][y] = "|" if num_adj_trees >= 3 else "."
            elif cur_iteration[x][y] == "|":
                next_iteration[x][y] = "#" if num_adj_lumberyards >= 3 else "|"
            elif cur_iteration[x][y] == "#":
                next_iteration[x][y] = "#" if num_adj_lumberyards >= 1 and num_adj_trees >= 1 else "."
            else:
                raise Exception("Invalid current square value: " + cur_iteration[x][y])

    if get_hash(next_iteration) in seen:
        break

    seen[get_hash(next_iteration)] = iteration

    cur_iteration = next_iteration
    iteration += 1

prev = seen[get_hash(next_iteration)]

num_remaining = (1000000000 - prev) % (iteration - prev)

for _ in range(num_remaining):
    next_iteration = [line[:] for line in cur_iteration]

    for y in range(grid_size):
        for x in range(grid_size):
            num_adj_trees = 0
            num_adj_lumberyards = 0

            for dy in range(-1, 2):
                for dx in range(-1, 2):
                    if (dx == 0 and dy == 0) or x + dx < 0 or y + dy < 0 or x + dx >= grid_size or y + dy >= grid_size:
                        continue

                    value = cur_iteration[x + dx][y + dy]
                    if value == "|":
                        num_adj_trees += 1
                    elif value == "#":
                        num_adj_lumberyards += 1

            if cur_iteration[x][y] == ".":
                next_iteration[x][y] = "|" if num_adj_trees >= 3 else "."
            elif cur_iteration[x][y] == "|":
                next_iteration[x][y] = "#" if num_adj_lumberyards >= 3 else "|"
            elif cur_iteration[x][y] == "#":
                next_iteration[x][y] = "#" if num_adj_lumberyards >= 1 and num_adj_trees >= 1 else "."
            else:
                raise Exception("Invalid current square value: " + cur_iteration[x][y])

    cur_iteration = next_iteration

num_trees = 0
num_lumberyards = 0

for y in range(grid_size):
    for x in range(grid_size):
        if cur_iteration[x][y] == "|":
            num_trees += 1
        elif cur_iteration[x][y] == "#":
            num_lumberyards += 1

print(num_trees * num_lumberyards)
